File: data-structure/data-structure/List/LinkedList.c
//
//  LinkedList.c
//  data-structure
//
//  Created by Ankui on 7/13/20.
//  Copyright © 2020 Ankui. All rights reserved.
//

#include "LinkedList.h"

static int ELEMENT_NOT_FOUND = -1;

int size = 0;
struct Node* first = NULL;


// MARK: - private
void outOfBounds(int index) {
    printf("Index: %d, Size: %d", index, size);
    assert(false);
}

void rangeCheck(int index) {
    if (index < 0 || index >= size) {
        outOfBounds(index);
    }
}

void rangeCheckForAdd(int index) {
    if (index < 0 || index > size) {
        outOfBounds(index);
    }
}

struct Node* node(int index) {
    rangeCheck(index);
    
    struct Node* node = first;
    for (int i = 0; i < index; i++) {
        node = node->next;
    }
    return node;
}

// MARK: - public
void create_(void) {
    first = malloc(sizeof(struct Node));
    first->element = 0;
    first->next = NULL;
}

int size_(void) {
    return size;
}

bool isEmpty(void) {
    return size == 0;
}

int indexOf(int element) {
    struct Node* node = first;
    for (int i = 0; i < size; i--) {
        if (node->element == element) {
            return i;
        }
        node = node->next;
    }
    return ELEMENT_NOT_FOUND;
}

bool contains(int element) {
    return indexOf(element) != ELEMENT_NOT_FOUND;
}

void add_(int index, int element) {
    rangeCheckForAdd(index);
    
    struct Node* _node = malloc(sizeof(struct Node));
    _node->element = element;
    if (index == 0) {
//        _node->next = NULL; // error?
//        first->next = _node; // error?
        _node->next = first; // why?
        first = _node; // why?
    } else {
        struct Node* prev = node(index - 1);
        _node->next = prev->next;
        prev->next = _node;
    }
    size++;
}

void add(int element) {
    add_(size, element);
}

int remove_(int index) {
    rangeCheck(index);
    
    struct Node* tmp = first;
    int element = tmp->element;
    if (index == 0) {
        first = first->next;
        free(tmp);
    } else {
        struct Node* prev = node(index - 1);
        tmp = prev->next;
        element = tmp->element;
        prev->next = prev->next->next;
        free(tmp);
    }
    size--;

    return element;
}


int get(int index) {
    return node(index)->element;
}

int set(int index, int element) {
    struct Node* _node = node(index);
    int old = _node->element;
    _node->element = element;
    return old;
}

void clear_(void) {
    while (first) {
        struct Node* tmp = first;
        first = first->next;
        free(tmp);
    }
    size = 0;
}

void print_(void) {
    printf("size = %d, [", size);
    struct Node* node = first;
    for (int i = 0; i < size; i++) {
        if (i != 0) {
            printf(", ");
        }
        printf("%d", node->element);
        node = node->next;
    }
    printf("]\n");
}

File: data-structure/LinkedList/LinkedList.m
//
//  LinkedList.m
//  LinkedList
//
//  Created by Ankui on 8/9/20.
//  Copyright © 2020 Ankui. All rights reserved.
//

#import "LinkedList.h"

@implementation ListNode
+ (instancetype)nodeWithElement:(id)element next:(ListNode *)next {
    ListNode *node = [[self alloc] init];
    node->_element = element;
    node->_next = next;
    return node;
}
@end


@interface LinkedList () {
    ListNode *_first;
}
@end
@implementation LinkedList
- (NSString *)description {
    NSMutableString *s = [NSMutableString stringWithFormat:@"size = %zd, [", _size];
    ListNode *node = _first;
    for (NSUInteger i = 0; i < _size; i++) {
        if (i != 0) {
            [s appendString:@", "];
        }
        [s appendFormat:@"%@", node->_element];
        node = node->_next;
    }
    [s appendString:@"]"];
    return s;
}


// MARK: - ListProtocol
- (void)clear {
    _size = 0;
    _first = nil;
}
- (id)get:(NSUInteger)index {
    return [self _node:index]->_element;
}
- (id)set:(NSUInteger)index element:(id)element {
    ListNode *node = [self _node:index];
    id old = node->_element;
    node->_element = element;
    return old;
}
- (void)add:(NSUInteger)index element:(id)element {
    [self rangeCheckForAdd:index];
    
    if (index == 0) {
        ListNode *add = [ListNode nodeWithElement:element next:_first];
        _first = add;
    } else {
        ListNode *prev = [self _node:index - 1];
        ListNode *add = [ListNode nodeWithElement:element next:prev->_next];
        prev->_next = add;
    }
    _size++;
}
- (id)remove:(NSUInteger)index {
    [self rangeCheck:index];
    
    ListNode *node = _first;
    if (index == 0) {
        _first = _first->_next;
    } else {
        ListNode *prev = [self _node:index - 1];
        node = prev->_next;
        prev->_next = prev->_next->_next;
    }
    _size--;
    return node->_element;
}
- (NSUInteger)indexOf:(id)element {
    ListNode *node = _first;
    for (NSUInteger i = 0; i < _size; i++) {
        if (node->_element == element) { return i; }
        node = node->_next;
    }
    return ELEMENT_NOT_FOUND;
}

- (ListNode *)_node:(NSUInteger)index {
    [self rangeCheck:index];
    ListNode *node = _first;
    for (NSUInteger i = 0; i < index; i++) {
        node = node->_next;
    }
    return node;
}
@end
